### DISCUSSION TITLE
Clojurescript Support for Browser

### DISCUSSION THREAD

**Discussion by jsmassa** [replikativ/datahike#429] [Ideas]
Should we support ClojureScript in the browser?

**Comment by chuckberrypi** [comment]
I don’t know if my situation is at all typical, but browser support would be huge for me. I’m not a professional developer, but I know enough that I’m able to write simple tools that would make my job much easier/more pleasant. The catch is that getting new applications approved on my firm’s machines is a pain, so writing programs to run in the browser is the only realistic way for me to build and use my own tools. Having something like this project to handle persistence in, say, indexeddb would be *huge* for my particular, admittedly-maybe-not-so-common use case. I’ve tried interfacing with it directly and have gotten stymied (again, not a pro…)

**Comment by garbados** [comment]
Hi! Any news here?

It seems there is an [outstanding PR](https://github.com/replikativ/datahike/pull/357) and the [branch](https://github.com/replikativ/datahike/tree/206-cljs-support) mentioned in the [Datahike in ClojureScript with IndexedDB support](https://lambdaforge.io/2021/03/03/datahike-clojurescript.html) blog post. It seems there has been a lot going on, now and then. I eagerly await this feature's arrival.

**Comment by whilo** [comment]
Hey @garbados! 

We have focused to get the JVM version running smoothly first, cover also native builds (which are harder to recover later once you have problematic dependencies) and deliver the distributed memory model with the default transaction interface of the new REST server with JSON support. @pkpkpk and I are now looking into porting the durability in the persistent sorted set to ClojureScript, which is the first step to complete Datahike ClojureScript support. Hopefully this can be done in the next few weeks and then we will have to plan for how to render the Datahike codebase async+sync.

What are you particularly interested in? Also more help would definitely speed things up.

**Reply by garbados** [reply]
I am trying to develop tooling for a client and server to replicate data, much like CouchDB, while supporting Datalog queries. Replication can happen over HTTP, WebRTC, or whatever, but it's important to be able to replicate data into the browser for persistent storage. As it stands, I'm inclined to develop my own tooling from the ground up, as I need to implement a specific tree structure (Prolly trees!) to support an efficient replication API. However, I was wondering if Datahike can accomplish the same thing, and save me the trouble of implementing a Datalog query engine. So, ClojureScript support is a necessary prerequisite for me. That said, I don't really know if Datahike is a good fit for my replication requirements.

**Reply by whilo** [reply]
That is interesting! Can you describe your replication requirements in a bit more detail? I.e. how would you store things and how to you expect them to be replicated?

**Reply by garbados** [reply]
So as pretext, I'm mostly familiar with CouchDB's *revision-style* replication model, which enables complex conflict resolution and replication in logarithmic time, but it also applies a considerable data model to the stored data. App devs must add logic and flows to their applications to asynchronously check for and raise conflicts, which only the user may know how to resolve. That introduces another UI headache entirely.

The thing is that most apps don't need the specific guarantees of revision-style replication. They want simple push/pull/sync/merge to deliver a good-enough multi-device experience, or they want something custom where replicated changes are handled by application logic. There haven't been a lot of great alternatives for these use-cases, leaving offline-first developers monopolized by an ecosystem with heavy assertions and disorganized management.

In recent years, the *Prolly tree* datastructure has [gained notoriety](https://docs.canvas.xyz/blog/2023-05-04-merklizing-the-key-value-store.html) as it supports the typical methods of a key/value store, such as CRUD and range queries, as well as replication in logarithmic time. It does this by "merklizing" a tree, where leaf nodes contain key/value pairs, and internal nodes contain the hashes of all their immediate descendant nodes. Thus, two trees contain all the same entries if their root nodes have the same hash. Trees synchronize -- or, really, just generate a stream of diffs -- by examining only those nodes that differ. That's how it gets the job done in logarithmic time. That stream of diffs can then either be handled by out-of-the-box sync functions or custom logic, while the tree itself can be used as a key/value store.

With all that said, I don't *need* to use a Prolly tree. As long as I can get efficient replication of database state between environments, specifically between a web server and a browser peer, then I'll be happy to build on top of that. For comparison, currently, I use PouchDB to store data in IndexedDB, while a proxy controls replicating with a locked-down CouchDB installation. Data is stored both on the browser and on the server, as the purpose of replication is to maintain multiple copies efficiently. PouchDB provides JavaScript indices, but it may prove self-evident that Datalog is a superior querying solution.

I have big plans for apps down the road, and this tooling is only the beginning... so, I'm hoping something like Datahike can save me some of that effort :sweat_smile:

**Reply by whilo** [reply]
Interesting! replikativ has fully merkelized data structures with delta fetching since its creation in 2013 through https://github.com/replikativ/hasch (which I created for this purpose) and when I picked up the hitchhiker-tree in 2017 I did it to do this logarithmic time synchronization between peers, but then got sidetracked to first make Datahike work (because its value proposition with a convenient query language is more obvious). It makes a lot of sense to me, I think I should have written more blog posts and maybe even papers about what I did, it was kind of obvious to me (I had touched on it briefly in this blog post from 2019 https://github.com/datopia/pixyll/blob/master/_posts/2018-11-03-hitchhiker-tree.md#replication). The hitchhiker-tree, and now the durable persistent sorted set, have exactly the key-value store functionality you are talking about and can be transferred with structural deltas.

It would be super cool if you would write the synchronization code, it should not be hard. 1. You get sent a root from another peer and then have to fetch all its children that you don't have, 2. and then do it recursively over these children until there is nothing left to fetch. Done. Ideally this should happen asynchronously and not block threads, which I do with core.async and https://github.com/replikativ/konserve.

The persistent-sorted-set has a faster read iterator speed than the hitchhiker-tree had and the hitchhiker-tree had lazy upserts, which are non-trivial to make performant. Having said this, there is also the opportunity to make data structures of https://github.com/lacuna/bifurcan durable if this is a better fit for you.

**Reply by whilo** [reply]
Note also that if the underlying store supports efficient delta transmission, such as rsync or datsync for the filesystem, https://lambdaforge.io/2019/12/08/replicate-datahike-wherever-you-go.html, then you can just use that, since the tree delta is projected one-to-one into the store through konserve and will be picked up efficiently by the transport mechanism.

We just have not done this for the web yet.

**Reply by garbados** [reply]
That's great news to hear! I'll peruse these links when I have some time. I'd love to contribute as I'm able! Point me where to start and I'll dive right in. I've written sync code for Prolly trees so writing it for your datastructure probably won't be too different.

**Reply by whilo** [reply]
Sorry for the late reply, Christmas break always takes some time meeting friends and family. One take on this would be to implement a Datahike datatype in replikativ (which would not be a CRDT) and then use its p2p pub-sub bus to distribute it to all peers (jvm or js). As long as there is only one writer in the system that would already work. The main work needed there is to update the replikativ repos and maybe improve some of the logic. Another take would be to use another pub-sub mechanism such as the dat protocol and just publish all the tree nodes and DB records that get created there. Which systems have you worked with? @garbados

In any case we ideally would stream the tree fragment deltas and then update the DB record. This would allow apps like logseq to automatically retrieve new Datahike/DataScript frontend instances. @tiensonqin and @logseq-cldwalker have worked already on an extension of the persistent-sorted-set storage protocol @tonsky and I developed here https://github.com/tonsky/persistent-sorted-set/pull/14. 

I am also happy to move to a chat medium and/or have a call about this.

**Reply by garbados** [reply]
No worries, it's the holidaze. I'm glad you got some time for you and yours.

I've worked with Dat, and particularly Hypercore which underpins it, as well as IPFS, SSB, and of course the venerable CouchDB. I've been around the block, but Datalog is more or less new to me. I tend to dislike Dat/Hypercore for pub-sub because it relies on an append-only log, which precludes hard deletes as they would compromise the chain of hashes that prove the log's authenticity. That's an unworkable caveat for the sorts of social apps that I build, where people generally expect to be able to tell the system to forget about their bad posts / mis-posts / unflattering selfies. The way I think I want to handle this is to construct a version of the underlying tree in IndexedDB, so we can do replication over WebRTC or HTTP or whatever; by having a feature-complete tree, a browser app can act as a fully autonomous writer on the tree, thus enabling offline-first apps. However, by making client devices into autonomous peers, we are forced to treat each device as a writer, which precludes the simplifying one-writer assumption. CouchDB solves this with document versioning, so peers can know which updates came before or after a specific document version, and can generate *conflicts* when disagreements happen, but this complicates the underlying data model and may thus be unworkable. Anyway, it sounds like there's some architecture details still to hammer out.

If I understand you, replikativ is the place to begin this work. I'll check out the persistent-sorted-set work you linked. Do you have any info to impart about the ClojureScript efforts on DataHike itself? There have been a few, but I don't have mastery over the source yet so what I can glean from commit diffs is limited. Any clarity you can provide will really help me situate myself in this work.

**Reply by whilo** [reply]
If you replicate the persistent-sorted-set then *both* Datahike and DataScript will effectively get replication. Datahike has not recently been compiled for ClojureScript, but assuming a synchronous execution (as is done in https://github.com/tonsky/persistent-sorted-set/pull/14) fixing Datahike cljs support would mostly be working through relatively simple compiler errors. This effort is much easier than porting the code base to support asynchronous IO, which is what we did for the ClojureScript port that is now stale, https://github.com/replikativ/datahike/tree/206-cljs-support. The latter has benefits, in particular it allows to lazily load index fragments from large remote DBs instead of always having to fetch fragments first, but the drawback being asynchronous execution which can be slower and less nice to handle in JS. 

You have covered a nice set of deeply related technologies and I  see your dat problem. There are potential fixes for that unless you cannot rely on physical clocks or don't know all replicas through vector clocks. But handling multiple writers/conflicts is not easy in general anyway as it can induce unintuitive app behaviour when things get undone asynchronously. Regarding social apps with replikativ, we had this demo Twitter-like app https://www.topiq.es/. It is not very polished, but it also used DataScript replication, in this case of the transaction log with CDVCS instead of the index. 

How can I clarify further?

**Reply by garbados** [reply]
These are excellent leads, thank you for teeing me up like this. Give me a minute to read up and chew on the problem surface, then I'll probably have more questions.

**Reply by whilo** [reply]
@garbados Lmk if you like to have a call to discuss the implementation. It should not be very difficult to sync the persistent sorted set.

**Reply by garbados** [reply]
That would be great. I'm still chewing through compilation bugs on my off-hours, but it would be good to sync up soon. Uh, how do you want to schedule something? Should I drop a line somewhere for DMs?

**Reply by whilo** [reply]
Yes, that would be great. You can also ping me on the Clojurians slack if you like.
